runtime/common/src/chain_ext.rs

```diff
@@ -193,4 +184,40 @@ impl From<dia::CoinInfo> for CoinInfo {
 			price: coin_info.price,
 		}
 	}
-}
\ No newline at end of file
+}
+
+pub struct DecodeByteReader {
+	remaining_len: usize,
+	vec: Vec<u8>,
+}
+impl DecodeByteReader {
+	pub fn new(vec: Vec<u8>) -> Self {
+		Self { remaining_len: vec.len(), vec }
+	}
+}
+impl Input for DecodeByteReader {
+	fn remaining_len(&mut self) -> Result<Option<usize>, codec::Error> {
+		Ok(Some(self.remaining_len))
+	}
+	fn read(&mut self, into: &mut [u8]) -> Result<(), codec::Error> {
```

Comment:
Having a doc comment what exactly this function does would be helpful. If I understand correctly it reads the contents of itself into another array, padding with `0`s if the `into` array is longer than its own content?

Reply:
Doc comments here are passed down from the Input trait. But I added a regular comment indicating that as well as a short tldr; about what it is for.

Reply:
I agree however, that it is not clear why this struct and trait implementation is here in the first place. Why is it required?

Reply:
It is explained in the comment I just added. T::decode() requires as input a type that implements the codec::Input trait which it uses to read the data it will decode. (In this case the Vec<u8> which is contained in the struct)

Reply:
@ashneverdawn This is still not clear to me: the comment describes the what but not the why. Why would we need to provide our own custom decoder logic and why would it look like this?

Reply:
I thought this was required, because the decode function would not accept the `Vec<u8>` directly. It expected a type that implements this Input trait (which is not implemented for `Vec<u8>`):
https://github.com/paritytech/parity-scale-codec/blob/52c9871cf8c71163afb8845589df63ab6c0e10a6/src/codec.rs#L65-L108
 
But now looking at this again, I see there actually exists an implementation for [u8]
So, I just pushed a commit removing my Vec<u8> implementation and instead calling `.as_slice()` on the `Vec<u8>`